R/estimateRichness.R

```diff
@@ -0,0 +1,300 @@
+#'                                index="observed",
+#'                                name = c("index3", "index4"))
+#' }
+#' # Shows all indices
+#' colData(esophagus)
+#' # Deletes all indices
+#' colData(esophagus) <- NULL
+#'
+#' # Calculates all indices
+#' esophagus <- estimateRichness(esophagus)
+#' # Shows all indices
+#' colData(esophagus)
+NULL
+
+
+#' @rdname estimateRichness
+#' @export
+setGeneric("estimateRichness",signature = c("x"),
+        function(x, abund_values = "counts",
+                    index = c("observed", "Chao1", "ACE", "Hill"),
+                    name = index,
+                    detection = 0,
+                    ...,
+                    BPPARAM = SerialParam())
+                    standardGeneric("estimateRichness"))
+
+
+
+#' @rdname estimateRichness
+#' @export
+setMethod("estimateRichness", signature = c(x = "SummarizedExperiment"),
+    function(x,
+            abund_values = "counts",
+            index = c("observed", "Chao1", "ACE", "Hill"),
+            name = index,
+            detection = 0,
+            ...,
+            BPPARAM = SerialParam()){
+        # Input check
+        # Check abund_values
+        .check_abund_values(abund_values, x)
+        # Check indices
+        index <- match.arg(index, several.ok = TRUE)
+        if(!.is_non_empty_character(name) || length(name) != length(index)){
+            stop("'name' must be a non-empty character value and have the ",
+                "same length than 'index'.",
+                call. = FALSE)
+        }
+
+        # Calculates richness indices
+        richness <- BiocParallel::bplapply(index,
+                                            FUN = .get_richness_values,
+                                            assay = assay(x,abund_values),
+                                            detection = detection,
+                                            BPPARAM = BPPARAM)
+        # Add richness indices to colData
+        .add_richness_values_to_colData(x, richness, name)
+    }
+)
+
+
+.get_observed <- function(x, detection = 0, ...){
+    # vegan::estimateR(t(x))["S.obs",]
+    colSums(x > detection)
+}
+
+.get_chao1 <- function(x, ...){
+    ans <- t(vegan::estimateR(t(x))[c("S.chao1","se.chao1"),])
+    colnames(ans) <- c("",".se")
+    ans
+}
+
+.get_ACE <- function(x, ...){
+    ans <- t(vegan::estimateR(t(x))[c("S.ACE","se.ACE"),])
+    colnames(ans) <- c("",".se")
+    ans
+}
+
+.get_hill <- function(x, ...){
+
+    # Exponent of Shannon diversity
+    exp(vegan::diversity(t(x), index="shannon"))
+
+}
+
+.calc_richness <- function(mat, detection, index){
+
+    ans <- lapply(mapply(function(i,j,x){x[i,j]},
+                        i = idx,
+                        j = seq_len(ncol(mat)),
+                        MoreArgs = list(x = mat),
+                        SIMPLIFY = FALSE),
+                        sum)
+    ans <- unlist(ans)
+
+    # Adds sample names to the table
+    names(ans) <- colnames(mat)
+    ans
+}
```

Comment:
the index argument is not used. and can be removed. It also doesn't really make sense in context of the subsequent function